### PR TITLE
Update TransactionFailureReason

### PR DESCRIPTION
**Pull request by thibault-martinez**
https://github.com/iotaledger/iota.go/pull/692

### CODE REVIEW

File: sdk/src/types/block/semantic/unlock.rs [hunk 1]
```diff
@@ -13,58 +13,63 @@ impl SemanticValidationContext<'_> {
...
 
-                if signature
-                    .is_valid(self.transaction_signing_hash.as_ref(), ed25519_address)
-                    .is_err()
-                {
-                    return Err(TransactionFailureReason::UnlockSignatureInvalid);
-                }
+                    if signature
+                        .is_valid(self.transaction_signing_hash.as_ref(), ed25519_address)
+                        .is_err()
```
Comment (DaughterOfMars) [inline sdk/src/types/block/semantic/unlock.rs]: hmm, weird. Does `is_valid` also return a bool we should check?

Comment (thibault-martinez) [reply sdk/src/types/block/semantic/unlock.rs]: `pub fn is_valid(&self, message: &[u8], address: &Ed25519Address) -> Result<(), Error>`

Comment (thibault-martinez) [reply sdk/src/types/block/semantic/unlock.rs]: Can fail for multiple reasons

Comment (DaughterOfMars) [reply sdk/src/types/block/semantic/unlock.rs]: probably shouldn't be called `is_valid` then, unless this isn't in our lib?

File: sdk/src/types/block/semantic/unlock.rs [hunk 2]
```diff
@@ -13,58 +13,63 @@ impl SemanticValidationContext<'_> {
...
-            }
-            (Address::Account(account_address), Unlock::Account(unlock)) => {
-                // PANIC: indexing is fine as it is already syntactically verified that indexes reference below.
-                if let (output_id, Output::Account(account_output)) = self.inputs[unlock.index() as usize] {
-                    if &account_output.account_id_non_null(output_id) != account_address.account_id() {
-                        // TODO https://github.com/iotaledger/iota-sdk/issues/1954
-                        return Err(TransactionFailureReason::SemanticValidationFailed);
-                    }
+                Unlock::Reference(_) => {
+                    // TODO actually check that it was unlocked by the same signature.
```
Comment (Thoralf-M) [inline sdk/src/types/block/semantic/unlock.rs:34]: I know it was there before already, but what does this mean?

File: sdk/src/types/block/semantic/unlock.rs (hunk 1 above)
Comment (thibault-martinez) [reply sdk/src/types/block/semantic/unlock.rs]: What would would you call it?

File: sdk/src/types/block/semantic/unlock.rs [hunk 3]
```diff
@@ -13,58 +13,63 @@ impl SemanticValidationContext<'_> {
...
-                    }
-                    if !self.unlocked_addresses.contains(address) {
-                        // TODO https://github.com/iotaledger/iota-sdk/issues/1954
-                        return Err(TransactionFailureReason::SemanticValidationFailed);
+            (Address::Nft(nft_address), unlock) => {
+                if let Unlock::Nft(unlock) = unlock {
+                    // PANIC: indexing is fine as it is already syntactically verified that indexes reference below.
+                    if let (output_id, Output::Nft(nft_output)) = self.inputs[unlock.index() as usize] {
+                        if &nft_output.nft_id_non_null(output_id) != nft_address.nft_id() {
+                            return Err(TransactionFailureReason::ChainAddressUnlockInvalid);
```
Comment (Thoralf-M) [inline sdk/src/types/block/semantic/unlock.rs:63]: This is really the correct error here? Same for accounts

File: sdk/src/types/block/semantic/unlock.rs (hunk 1 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/semantic/unlock.rs]: `validate` probably

File: sdk/src/types/block/semantic/unlock.rs (hunk 3 above)
Comment (Thoralf-M) [reply sdk/src/types/block/semantic/unlock.rs:63]: Ah I guess yes

Comment (thibault-martinez) [reply sdk/src/types/block/semantic/unlock.rs:63]: I would say so, yes

File: sdk/src/types/block/semantic/unlock.rs (hunk 1 above)
Comment (thibault-martinez) [reply sdk/src/types/block/semantic/unlock.rs]: Ok I can change it

Comment (thibault-martinez) [reply sdk/src/types/block/semantic/unlock.rs]: Done

**Review by DaughterOfMars — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

File: sdk/src/types/block/semantic/unlock.rs (hunk 2 above)
Comment (Thoralf-M) [reply sdk/src/types/block/semantic/unlock.rs:34]: The referenced signature could be different, so we need to get it by index and compare the hashed public key with the pubkeyhash of the address

**Pull request merged**